libraries/aws/shadow/src/shadow.c

```diff
@@ -0,0 +1,550 @@
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DELTA
+    };
+
+    /* Lookup table for Shadow message string length. */
+    static const uint16_t pMessageStringsLength[ ShadowMessageTypeMaxNum ] =
+    {
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DELTA
+    };
+
+    /* Lookup table for Shadow message types. */
+    static const ShadowMessageType_t pMessageTypes[ ShadowMessageTypeMaxNum ] =
+    {
+        ShadowMessageTypeGetAccepted ,
+        ShadowMessageTypeGetRejected,
+        ShadowMessageTypeDeleteAccepted,
+        ShadowMessageTypeDeleteRejected,
+        ShadowMessageTypeUpdateAccepted,
+        ShadowMessageTypeUpdateRejected,
+        ShadowMessageTypeUpdateDocument,
+        ShadowMessageTypeUpdateDelta
+    };
+
+    for( ; index < ( uint32_t ) ( sizeof( pMessageStrings ) / sizeof( pMessageStrings[0] ) ); index++ )
+    {
+        returnStatus = containsSubString( pString,
+                                          stringLength,
+                                          pMessageStrings[ index ],
+                                          pMessageStringsLength[ index ] );
+
+        /* If the operation string matches, there must not be any other extra
+         * character remaining in the string. */
+        if( returnStatus == SHADOW_STATUS_SUCCESS )
+        {
+            if( stringLength != pMessageStringsLength[ index ] )
+            {
+                returnStatus = SHADOW_STATUS_FAIL;
+            }
+            else
+            {
+                * pMessageType = pMessageTypes[ index ];
+                break;
+            }
+        }
+    }
+
+    return returnStatus;
+}
+/*-----------------------------------------------------------*/
+
+ShadowStatus_t Shadow_MatchTopic( const char * pTopic,
+                                  uint16_t topicLength,
+                                  ShadowMessageType_t * pMessageType,
+                                  const char ** pThingName,
+                                  uint16_t * pThingNameLength )
+{
+    uint16_t consumedTopicLength = 0U;
+    ShadowStatus_t shadowStatus = SHADOW_STATUS_SUCCESS;
+
+    if( ( pTopic == NULL ) ||
+        ( topicLength == 0U ) ||
+        ( pMessageType == NULL ) ||
+        ( pThingName == NULL ) ||
+        ( pThingNameLength == NULL ) )
+    {
+        shadowStatus = SHADOW_STATUS_BAD_PARAMETER;
+        LogError( ( "Invalid input parameters" ) );
```

Comment:
We should probably log the passed parameters so we can see which parameter is invalid

Reply:
will add it in next commit, thank you

Reply:
fixed it in commit [a9b690e](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1081/commits/a9b690e63f7c1b3fbe90b7328fabb96b7f119eb7)

resolving this comment